docs/source/upgrade.rst

```diff
@@ -0,0 +1,759 @@
+The easiest fix is setting the ``chainer.config.keep_graph_on_report`` flag to ``True``.
+*Note that this fix will cancel the improvement on the memory consumption made in Chainer v2.*
+More memory-efficient fix is to dump the graph without using an extension, e.g. by customizing the loss function or the updater.
+
+Here is the background of this change.
+In Chainer v2, :ref:`the Reporter copies reported variables with purging the computational graph by default. <upgrade-reporter-purge-variable>`
+On the other hand, the :func:`~training.extensions.dump_graph` extension requires the computational graph reachable from the reported variable.
+In order to make the graph available, the :func:`~training.extensions.dump_graph` extension turns on the ``chainer.config.keep_graph_on_report`` flag at its initializer (i.e., it turns on the graph before entering the training loop).
+Since we also wanted to achieve the memory efficiency, the :func:`~training.extensions.dump_graph` extension **turns off the flag after dumping the graph at its first invocation** (strictly speaking, it recovers the original value).
+As a result, the computational graph is not available from the second invocation.
+
+Since the :func:`~training.extensions.dump_graph` recovers the original flag value at its invocation, you can keep the graph dumped more than once by changing the original flag value.
+
+
+Reporter
+--------
+
+.. _upgrade-reporter-purge-variable:
+
+When a variable is reported, the variable is copied with the graph purged
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In Chainer v2, when a :class:`Variable` object is reported using :func:`report` function (or directly using :class:`Reporter`), a copy of the variable is made without preserving the computational graph.
+**If your code depends on the reachability of the computational graph from the reported variable, you have to update your code.**
+The easiest way to update your code is setting ``chainer.config.keep_graph_on_report`` to ``True``, then Chainer will keep the computational graph reachable from the reported variable.
+
+This change is made for the memory performance reasone; with this change, the memory used by the computational graph for training is immediately released before invoking extensions.
+Therefore, *changing the behavior by overwriting* ``chainer.config.keep_graph_on_report`` *may increase the memory consumption.*
+It may cause an out-of-memory error if the computational graph of the loss function consumes almost all the memory available in your environment and there is an extension that uses a certain amount of memory (e.g. :class:`~training.extensions.Evaluator`).
+
+Other utilities
+---------------
+
+Some obsolete classes and functions are removed
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following classes and functions are removed in Chainer v2.
+
+- ``chainer.Flag``
+- ``chainer.FunctionSet``
+- ``chainer.cuda.init``
```

Comment:
: it did nothing but calling `check_cuda_available`. Call the function directly if you want to